File: symfony/src/Service/Players/PlayerService.php
<?php

namespace App\Service\Players;

use App\Entity\Players\Player;
use App\Repository\Players\PlayerRepository;

class PlayerService
{
    public function __construct(
        private PlayerRepository $repository,
    ) {}

    public function create(array $data): Player
    {
        throw new \LogicException('Not implemented');
    }

    public function update(Player $entity, array $data): Player
    {
        throw new \LogicException('Not implemented');
    }

    public function promote(int $id): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Player not found: ' . $id);
        $result = $entity->promote();
        $this->repository->save($entity, flush: true);
        return $result;
    }

    public function demote(int $id): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Player not found: ' . $id);
        $result = $entity->demote();
        $this->repository->save($entity, flush: true);
        return $result;
    }

    public function recordWin(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Player not found: ' . $id);
        $entity->recordWin();
        $this->repository->save($entity, flush: true);
    }

    public function recordLoss(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Player not found: ' . $id);
        $entity->recordLoss();
        $this->repository->save($entity, flush: true);
    }

    public function winRate(int $id): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Player not found: ' . $id);
        $result = $entity->winRate();
        $this->repository->save($entity, flush: true);
        return $result;
    }

    public function verify(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Player not found: ' . $id);
        $entity->verify();
        $this->repository->save($entity, flush: true);
    }

    public function updateRating(int $id, $delta): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Player not found: ' . $id);
        $entity->updateRating($delta);
        $this->repository->save($entity, flush: true);
    }
}

File: rails/app/services/players/player_service.rb
module Players
  class PlayerService
    def initialize(repository = Player)
      @repository = repository
    end

    def create(attributes)
      raise NotImplementedError, "#{self.class}#create not implemented"
    end

    def update(entity, attributes)
      raise NotImplementedError, "#{self.class}#update not implemented"
    end

    def promote(id)
      instance = Player.find(id)
      result = instance.promote()
      instance.save!
      result
    end

    def demote(id)
      instance = Player.find(id)
      result = instance.demote()
      instance.save!
      result
    end

    def record_win(id)
      instance = Player.find(id)
      instance.record_win()
      instance.save!
    end

    def record_loss(id)
      instance = Player.find(id)
      instance.record_loss()
      instance.save!
    end

    def win_rate(id)
      instance = Player.find(id)
      result = instance.win_rate()
      instance.save!
      result
    end

    def verify(id)
      instance = Player.find(id)
      instance.verify()
      instance.save!
    end

    def update_rating(id, delta)
      instance = Player.find(id)
      instance.update_rating(delta)
      instance.save!
    end
  end
end
